src/crypto-providers/signingValidation.ts

```diff
@@ -0,0 +1,232 @@
+  const numMintWitnesses = params.unsignedTxParsed.mint?.length || 0
+
+  if (numStakeWitnesses > 0 && (stakeSigningFiles.length === 0)) {
+    throw Error(Errors.MissingStakeSigningFileError)
+  }
+  if (numPoolColdWitnesses > 0 && (poolColdSigningFiles.length === 0)) {
+    throw Error(Errors.MissingPoolColdSigningFileError)
+  }
+  if (numMintWitnesses > 0 && (mintSigningFiles.length === 0)) {
+    throw Error(Errors.TooManyMintSigningFilesError)
+  }
+}
+
+const validatePoolRegistrationTx = (unsignedTxParsed: _UnsignedTxParsed) => {
+  _validateNonMultisigTx(unsignedTxParsed)
+
+  if (unsignedTxParsed.inputs.length === 0) {
+    throw Error(Errors.MissingInputError)
+  }
+  if (unsignedTxParsed.certificates.length !== 1) {
+    throw Error(Errors.MultipleCertificatesWithPoolRegError)
+  }
+  if (unsignedTxParsed.withdrawals.length > 0) {
+    throw Error(Errors.WithdrawalIncludedWithPoolRegError)
+  }
+}
+
+const validatePoolOwnerWitnesses = (params: SigningParameters) => {
+  const {
+    paymentSigningFiles, stakeSigningFiles, poolColdSigningFiles, mintSigningFiles, multisigSigningFiles,
+  } = filterSigningFiles(params.hwSigningFileData)
+
+  if (paymentSigningFiles.length > 0) {
+    throw Error(Errors.TooManyPaymentFilesWithPoolRegError)
+  }
+  if (stakeSigningFiles.length === 0) {
+    throw Error(Errors.MissingStakeSigningFileError)
+  }
+  if (stakeSigningFiles.length > 1) {
+    throw Error(Errors.TooManyStakeSigningFilesError)
+  }
+  if (poolColdSigningFiles.length > 0) {
+    throw Error(Errors.TooManyPoolColdSigningFilesError)
+  }
+  if (mintSigningFiles.length > 0) {
+    throw Error(Errors.TooManyMintSigningFilesError)
+  }
+  if (multisigSigningFiles.length > 0) {
+    throw Error(Errors.TooManyMultisigSigningFilesError)
+  }
+}
+
+const validatePoolOperatorWitnesses = (params: SigningParameters) => {
+  const {
+    stakeSigningFiles, poolColdSigningFiles, multisigSigningFiles,
+  } = filterSigningFiles(params.hwSigningFileData)
+
+  if (poolColdSigningFiles.length === 0) {
+    throw Error(Errors.MissingPoolColdSigningFileError)
+  }
+  if (stakeSigningFiles.length > 0) {
+    throw Error(Errors.TooManyStakeSigningFilesError)
+  }
+  if (multisigSigningFiles.length > 0) {
+    throw Error(Errors.TooManyMultisigSigningFilesError)
+  }
+}
+
+const validateMultisigTx = (unsignedTxParsed: _UnsignedTxParsed) => {
+  // We require multisig transactions to use certificates and withdrawals
+  // only with ScriptHash stake credentials
+  const certificatesOk = certificatesWithStakeCredentials(unsignedTxParsed.certificates).every(
+    (cert) => cert.stakeCredential.type === StakeCredentialType.SCRIPT_HASH,
+  )
+  const withdrawalsOk = unsignedTxParsed.withdrawals.every(
+    (withdrawal) => withdrawal.stakeCredential.type === StakeCredentialType.SCRIPT_HASH,
+  )
+  if (!certificatesOk || !withdrawalsOk) {
+    throw Error(Errors.NonScriptStakeCredentialInMultisigTx)
+  }
+}
```

Comment:
This is essentially the same function as `_validateNonMultisigTx`. They differ only in the error thrown and the stake credential type. Could we perhaps make one more generic function and then use it in both places?
Something like `validateStakeCredentialsAreAllOfType(stakeCredentailType: StakeCredentialType, unsignedTxParsed: _UnsignedTxParsed)`

Reply:
Fixed.